src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
+ * it has been synchronised.
+ *
+ * Returns a pointer to a configured libipt block decoder or NULL on error.
+ *
+ * XXX This function assumes that we are decoding a trace for the current CPU
+ * and for the current process.
+ */
+void *
+perf_pt_start_decode(void *buf, uint64_t len, int vdso_fd, char *vdso_filename,
+                     int *decoder_status) {
+    bool failing = false;
+
+    // Make a block decoder configuration.
     struct pt_config config;
-    pt_blk_alloc_decoder(&config);
+    memset(&config, 0, sizeof(config));
+    config.size = sizeof(config);
+    config.begin = buf;
+    config.end = buf + len;
+    config.flags.variant.block.end_on_call = 1;
+    config.flags.variant.block.end_on_jump = 1;
+
+    // Decode for the current CPU.
+    struct pt_block_decoder *decoder = NULL;
+    if (pt_cpu_read(&config.cpu) != 0) {
+        failing = true;
+        goto clean;
+    }
+
+    // Work around CPU bugs.
+    if (config.cpu.vendor) {
+        if (pt_cpu_errata(&config.errata, &config.cpu) < 0) {
+            failing = true;
+            goto clean;
+        }
+    }
+
+    // Instantiate a decoder.
+    decoder = pt_blk_alloc_decoder(&config);
+    if (decoder == NULL) {
+        failing = true;
+        goto clean;
+    }
+
+    // Sync the decoder.
+    *decoder_status = pt_blk_sync_forward(decoder);
+    if (*decoder_status == -pte_eos) {
+        // There were no blocks in the stream. The user will find out on next
+        // call to perf_pt_next_block().
+        goto clean;
+    } else if (*decoder_status < 0) {
+        failing = true;
+        goto clean;
+    }
+
+    // Build and load a memory image from which to recover control flow.
+    struct pt_image *image = pt_image_alloc(NULL);
+    if (image == NULL) {
+        failing = true;
+        goto clean;
+    }
+
+    struct load_self_image_args load_args = {image, vdso_fd, vdso_filename};
+    if (!load_self_image(&load_args)) {
+        failing = true;
+        goto clean;
+    }
+
+    if (pt_blk_set_image(decoder, image) < 0) {
+        failing = true;
+        goto clean;
+    }
+
+clean:
+    if (failing) {
+        pt_blk_free_decoder(decoder);
+        return NULL;
+    }
+    return decoder;
+}
+
+/*
+ * Updates `*addr' with the start address of the next block in the instruction
+ * stream.
+ *
+ * A start address of 0 indicates that the end of the instruction stream has
+ * been reached.
+ *
+ * `*decoder_status` will be updated with the new decoder status after the operation.
+ *
+ * Returns true on success or false otherwise. Upon failure, the start address
+ * is undefined.
+ */
+bool
+perf_pt_next_block(struct pt_block_decoder *decoder, int *decoder_status,
+                   uint64_t *addr) {
+    // If there are events pending, drain them first.
+    *decoder_status = drain_events(decoder, *decoder_status);
+    if (*decoder_status < 0) {
+        // Error.
+        return false;
+    } else if (*decoder_status & pts_eos) {
```

Comment:
What's `pts_eos`? I'm not sure I understand the second and third branches of this `if` statement TBH.

Reply:
`pts_eos` means "PT status: end of stream". In other words, there are no more blocks to iterate over. We return 0 which prompts the Rust iterator to return `None`.

The third case handles statuses that I've not yet considered. There's a list of possibilities here:
https://github.com/01org/processor-trace/blob/24982c1a6fce48f1e416461d42899805f74fbb26/libipt/include/intel-pt.h.in#L1064-L1074

* `pts_event_pending` should not occur, since this very call, by definition, should be draining the events from the stream.
* I don't think `pts_ip_suppressed` can happen as this call doesn't fetch any blocks.

None the less, I'd like to know if we get here. Would it be better as an assertion?

Reply:
Yes, maybe an `assert` would be better because at least we'd know if they *can* happen.